students/EricAdams/session06/test_mailroom.py

```diff
@@ -0,0 +1,221 @@
+    '''
+    result = mailroom.goodbye()
+    assert result == ('uit')
+
+
+def test_goodbye_pass():
+    '''
+    Provide a valid input, ('quit')
+    '''
+    result = mailroom.goodbye()
+    assert result == ('quit')
+
+
+def test_print_donors(capsys):
+    '''
+    This test captures stdout and asserts that there is an entry from
+    the donor_db in the screen output.
+    Saw this on stackoverflow.com.... seems to resolve the issue
+    of testing screen output.
+    '''
+    mailroom.print_donors()
+    out, err = capsys.readouterr()
+    sys.stdout.write(out)
+    donor = list(mailroom.donor_db.keys())
+    assert out.find(donor[0])
+
+
+def test_find_donor_key():
+    '''
+    Input a valid user from the mailroom.donor_db
+    '''
+    result = mailroom.find_donor('William Gates, III')
+    assert result == 'William Gates, III'
+
+
+def test_find_donor_nonexistant():
+    result = mailroom.find_donor('Eric Adams')
+    assert result is None
+
+
+def test_main_menu_selection_t():
+    '''
+    Pass the test by giving a user input of 't'
+    Simulate user input of 't'
+    :data - a text file
+            in the current directory, (test_main_menu_selection_t_inputs.txt),
+            is used as input data.
+            The value that the user would input is simulated
+            by making a one line entry into the file.
+            Example: for a user input of 't', enter the single
+            character 't' on the first line, (no quotes).
+
+    Saw this solution to the 'input' issue on stackoverflow.
+    Seems to work
+    '''
+    orig_stdin = sys.stdin
+    with open("test_main_menu_selection_t_inputs.txt") as sys.stdin:
+        result = mailroom.main_menu_selection()
+    sys.stdin = orig_stdin
+    assert result == "t"
+
+
+def test_gen_letter():
+    '''
+    Input a valid donor and assert that the letter is returned
+    '''
+    letter = dedent('''
+          Dear {}
+
+          Thank you for your very kind donation of ${:.2f}.
+          It will be put to very good use.
+
+                         Sincerely,
+                            -The Team
+          ''')
+    donor_list = list(mailroom.donor_db.keys())
+    donor = donor_list[0]
+    result = mailroom.gen_letter(donor)
+    assert result == letter.format(donor, mailroom.donor_db[donor][-1])
+
+
+def test_send_thank_you_list_menu(capsys):
+    '''
+    Simulate user input of 'list' and the 'menu', (to force a return)
+    :data - a text file
+            in the current directory, (test_send_thank_you_list_inputs.txt),
+            is used as input data.
+            The value that the user would input is simulated
+            by making a one entry into the file per user input.
+            Example: for a user input of 'list', enter 'list'
+            on the first line, (no quotes).
+            Enter 'menu' on the second line for the second user
+            input, (this also forces a return).
+    Modified from a solution to the 'input' issue on stackoverflow.
```

Comment:
nice job! 

This is called "monkey patching" and is a good way to test this kind of thing.

However, it's a bit sensitive to errors. If there is an error between re-assigning sys.stdin and putting it back, then yoursystem will be left in an ustable state.

What you are doing is "mocking" -- and the mock module gives you a way to do it that is more robust -- essentially handling the "make sure it's put back the way it was" part for you -- plus other features.

we'll get into that more when we get into advanced testing techniques in the next quarter.